### PR TITLE
Trm 20200529

### PR DESCRIPTION
**Pull request by TimMangan**
This PR contains the following:

- Fix for PsfLauncher crash due to incorrect string processing on working directory=""

- Fix for FRF for file accesses using "local device" syntax, i.e. "\\\\.\something". These are now ignored by the FRF.

- Added a new fixup module called "RegLegacyFixups".  Currently this supports only one type of registry fixup to solve issues with requests to open/create registry keys using permissions not allowed by MSIX, by changing the access permissions being requested.

- Added small debugging improvements to the unit tests.

### CODE REVIEW

**Comment by mattjreynolds** [comment]
@TimMangan I've been looking into how to resolve some registry issues with packaging 3rd party applications. Looks like you already have a solution - perfect timing, I'll test this on my packages and let you know how I get on.
Do you know if anyone has looked into possible fixup solutions for applications that want to write to HKLM? e.g redirecting into a location within HKCU

thanks,
Matt

**Comment by TimMangan** [comment]
@mattjreynolds Not at this time.  The PSF can't see the application hive directly, so we need either some Microsoft support or figure out where we want to redirect to.  

If we have to do it ourselves, I am currently favoring adding a HKCU key to the package to redirect to; the PSF would redirect to there and then MSIX would redirect to wherever the heck it does that.  But we'll just start small with this fix and then add to it with other fixup types within RegLegacy in the future.

**Comment by dhoehna** [comment]
@mattjreynolds That is an awesome idea actually.  One issue I see with that is if the application opens a second executable that expects something to be in HKLM. 

I don't know if it would be a problem. But I like the idea.  I'll keep that in my notes as a possible PSF fixup.

File: fixups/FileRedirectionFixup/GetPrivateProfileStringFixup.cpp [hunk 1]
```diff
@@ -23,15 +23,21 @@ DWORD __stdcall GetPrivateProfileStringFixup(
             DWORD GetPrivateProfileStringInstance = ++g_FileIntceptInstance;
             if constexpr (psf::is_ansi<CharT>)
             {
-                LogString(GetPrivateProfileStringInstance,L"GetPrivateProfileStringFixup for fileName", widen(fileName, CP_ACP).c_str());
-                LogString(GetPrivateProfileStringInstance,L" Section", widen_argument(appName).c_str());
-                LogString(GetPrivateProfileStringInstance,L" Key", widen_argument(keyName).c_str());
+                if (fileName != NULL)
+                    LogString(GetPrivateProfileStringInstance,L"GetPrivateProfileStringFixup for fileName", widen(fileName, CP_ACP).c_str());
```
Comment (dhoehna) [inline fixups/FileRedirectionFixup/GetPrivateProfileStringFixup.cpp:27]: Please include brackets, even on one line if and else blocks.

File: fixups/RegLegacyFixups/InitializeFixup.cpp [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+                Log("RegLegacyFixups: process spec\n");
+
+                Reg_Remediation_Spec specItem;
+
+                auto& specObject = spec.as_object();
+                Log("RegLegacyFixups: Have specObject\n");
+
+                auto type = specObject.get("type").as_string().wstring();
+                Log("RegLegacyFixups: have type");
+#if _DEBUG
```
Comment (dhoehna) [inline fixups/RegLegacyFixups/InitializeFixup.cpp:141]: Can you please move all the #IF _DEBUG calls to one method and call the logging function regardless if we are in debug or release.

File: PsfLauncher/main.cpp [hunk 3]
```diff
@@ -51,6 +53,11 @@ int __stdcall wWinMain(_In_ HINSTANCE, _In_opt_ HINSTANCE, _In_ PWSTR args, _In_
 int launcher_main(PCWSTR args, int cmdShow) noexcept try
 {
     Log("\tIn Launcher_main()");
+
+    //Log("\tWaiting for debugger to attach...");
```
Comment (dhoehna) [inline PsfLauncher/main.cpp:57]: Please remove commented out code if it is not needed.

**Comment by dhoehna** [comment]
> ��Microsoft (R) MakeAppx Tool

Please remove all the .txt files from the PR.

---
Refers to: tests/scenarios/Appx/ArchitectureTest.appx.makeappx.txt:1 in 1378054. [](commit_id = 13780548a05d071a1d58dc4d27cb594bd5463206, deletion_comment = False)

**Comment by dhoehna** [comment]
@TimMangan Went over the PR.  Added a few comments.  Please address them when you can.

File: fixups/FileRedirectionFixup/GetPrivateProfileStringFixup.cpp (hunk 1 above)
Comment (TimMangan) [reply fixups/FileRedirectionFixup/GetPrivateProfileStringFixup.cpp:27]: Updated.

File: fixups/RegLegacyFixups/InitializeFixup.cpp (hunk 2 above)
Comment (TimMangan) [reply fixups/RegLegacyFixups/InitializeFixup.cpp]: done.

File: fixups/FileRedirectionFixup/GetPrivateProfileStringFixup.cpp (hunk 1 above)
Comment (TimMangan) [reply fixups/FileRedirectionFixup/GetPrivateProfileStringFixup.cpp:27]: done

File: PsfLauncher/main.cpp (hunk 3 above)
Comment (TimMangan) [reply PsfLauncher/main.cpp]: done.

**Comment by TimMangan** [comment]
Removed logging txt files as well.  All requests are processed.

File: fixups/RegLegacyFixups/readme.md [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+		"dll":"RegLegacyFixups.dll",
+		"config": [
+			{
+				"type": "ModifyKeyAccess",
+				"remediation": [
+					{
+						"hive": "HKCU",
+						"patterns": [
+							"^Software\\\\Vendor.*"
+						]
```
Comment (mattjreynolds) [inline fixups/RegLegacyFixups/readme.md:70]: The trailing comma in missing in this example.
```suggestion
						],
```

File: fixups/RegLegacyFixups/readme.md [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+						"access": "Full2RW"
+					},
+										{
+						"hive": "HKLM",
+						"patterns": [
+							"^[Ss][Oo][Ff][Tt][Ww][Aa][Rr][Ee]\\\\Vendor.*"
+						]
```
Comment (mattjreynolds) [inline fixups/RegLegacyFixups/readme.md:77]: The trailing comma in missing in this example.
```suggestion
						],
```

Comment (TimMangan) [reply fixups/RegLegacyFixups/readme.md]: Good catch.  Added to the PT.

File: fixups/FileRedirectionFixup/GetPrivateProfileStringFixup.cpp (hunk 1 above)
Comment (dhoehna) [reply fixups/FileRedirectionFixup/GetPrivateProfileStringFixup.cpp:27]: Thank you.  I don't see the change in your PR?  Did you push the changes?  For line 26-31

---
In reply to: [439763757](https://github.com/microsoft/MSIX-PackageSupportFramework/pull/145#discussion_r439763757) [](ancestors = 439763757)

Comment (TimMangan) [reply fixups/FileRedirectionFixup/GetPrivateProfileStringFixup.cpp:28]: Sorry, I caught the ones below that and missed those 3.  Updated the PR for those lines.

**Pull request merged**